### DISCUSSION TITLE
Print warning early in pytest output

### DISCUSSION THREAD

**Discussion by JP-Ellis** [pytest-dev/pytest#13056] [Q&A] [answered]
I have a repository whereby the majority of the tests are independent and using `python-xdist` works great; however, I also have a specific directory containing a set of tests which must be executed in a specific order and therefore is incompatible with having multiple xdist workers.

For that directory containing these order-dependent tests, I want to raise a warning if multiple workers are detected. For example, something like adding this to the directory's `conftest.py`:

```python
@pytest.fixture(scope="session", autouse=True)
def check_xdist() -> None:
    if int(os.getenv("PYTEST_XDIST_WORKER_COUNT", "1")):
        warnings.warn(
            "Running the examples with multiple workers may cause issues. "
            "Consider running the examples with a single worker."
        )
```

The problem with this solution is that when the test inevitably fails, the warning is buried below all the error messages that (most likely) inevitably are generated.

I have considered replacing `warnings.warn` with `pytest.fail`, but this would prevent the test from running at all which is overkill and would prevent individual tests from running.

So my question is:

<b>

Can I ensure that the `warnings.warn` call (or equivalent) is printed early? E.g., to produce an output akin to:

```console
===================================== test session starts =====================================
platform darwin -- Python 3.13.1, pytest-8.3.4, pluggy-1.5.0
rootdir: /Users/joshua.ellis/src/pact-foundation/pact-python
configfile: pyproject.toml
plugins: asyncio-0.24.0, cov-6.0.0,nyio-4.7.0, bdd-8.1.0, xdist-3.6.1
asyncio: mode=Mode.STRICT, default_loop_scope=session
10 workers [18 items]     
WARNING: Detected multiple xdist workers. Examples will most likely fail.   <-- THIS
...EEEF...FF.F
```

</b>

EDIT: to be clear, I do _not_ want solutions which require addition command line arguments, as that would imply that the person calling pytest is already aware of the warning. I want the warning to be displayed when people are naively running pytest.

**Comment by dongfangtianyu** [comment]
> the warning is buried below all the error messages that (most likely) inevitably are generated.


Implemented by the built-in plugin `warnings`, you can use the `-p` to disable the plugin and use the `-s` to display warning content

```python
@pytest.fixture(scope='session', autouse=True)
def hello():
    warnings.warn('hello')

```

```
pytest -p no:warnings -s -n3
 ======================== test session starts ========================
platform win32 -- Python 3.12.0, pytest-8.3.4, pluggy-1.5.0
rootdir: C:\pytest834
configfile: pytest.ini
plugins: xdist-3.6.1
3 workers [3 items]     
C:\pytest834\conftest.py:8: UserWarning: hello
  warnings.warn('hello')
C:\pytest834\conftest.py:8: UserWarning: hello
  warnings.warn('hello')
C:\pytest834\conftest.py:8: UserWarning: hello
  warnings.warn('hello')
...
 ======================== 3 passed in 0.70s ========================
```

**Reply by JP-Ellis** [reply]
The problem with that solution is that it requires prior knowledge of the warning to know to add `-p no:warnings` and `-s`. By that stage, they already have seen the warning and should not be adding `-n3`.

Imagine instead of scenario where `PYTEST_ADDOPTS` has been set, and pytest is executed as part of an IDE where one does not see/control the arguments as easily. I want the warning to appear _then_.

**Reply by dongfangtianyu** [reply]
Don't need to know anything in advance, just use command parameters, configuration files, or `PYTEST ADDOPTS`, and use `-p` and `- s` to take effect. If everything is uncontrollable, then the outcome will not change

**Reply by JP-Ellis** [reply]
Are you trying to say that without adding `-s` and `-p no:warnings`, what I'm asking is impossible?

I guess an alternative question would be: how can I adjust/augment the information that is printed at the start of a pytest run? Is it possible to hook in to pytest's print-to-terminal functionality (e.g., the way it displays information at the start, or progress during runs)?

**Reply by dongfangtianyu** [reply]
Three ideas:
1. Use the method provided by pytest (through configuration or parameters)
2. Change pytest (through hook or modify source code)
3. Execute custom python code before pytest

Depends on what you can control. If you can't control parameters, configuration files, or environment variables, but can modify `conftest.py`, or you can try hook

**Comment by JP-Ellis** [comment, marked as answer]
The solution ultimately was to use [a hook](https://docs.pytest.org/en/stable/how-to/writing_hook_functions.html#declaring-new-hooks). I can't speak for _all_ hooks, but it would appear that PyTest's stdout and stderr capturing are not yet enabled, thereby making warnings like I have above print to the console.

In my specific case of `pytest-xdist`, they conveniently expose the hook `pytest_xdist_setupnodes` which I can inspect to verify whether multiple workers are being used:

#### `conftest.py`

```python
def pytest_xdist_setupnodes(
    config: pytest.Config,
    specs: Sequence[execnet.XSpec],
) -> None:
    if len(specs) > 1:
        sys.stderr.write("\n")
        warnings.warn(
            "Running the examples with multiple workers may cause issues. "
            "Consider running the examples with a single worker by setting "
            "`--numprocesses=1` or using `hatch run example`.",
            stacklevel=1,
        )
```
